src/main/java/org/wildfly/extension/elytron/FileSystemRealmDefinition.java

```diff
@@ -0,0 +1,168 @@
+                    .build();
+
+    static final SimpleAttributeDefinition LEVELS =
+            new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.LEVELS, ModelType.INT, true)
+                    .setDefaultValue(new ModelNode(2))
+                    .build();
+
+    static final SimpleAttributeDefinition NAME_REWRITER =
+            new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.NAME_REWRITER, ModelType.STRING)
+                    .setAttributeGroup(ElytronDescriptionConstants.NAME_REWRITER)
+                    .setXmlName(ElytronDescriptionConstants.NAME)
+                    .setCapabilityReference(NAME_REWRITER_CAPABILITY, SECURITY_REALM_CAPABILITY, true)
+                    .setAllowNull(true)
+                    .build();
+
+    private static final AttributeDefinition[] ATTRIBUTES =
+            new AttributeDefinition[]{PATH, RELATIVE_TO, LEVELS, NAME_REWRITER};
+
+    private static final AbstractAddStepHandler ADD = new RealmAddHandler();
+    private static final OperationStepHandler REMOVE = new RealmRemoveHandler(ADD);
+
+    FileSystemRealmDefinition() {
+        super(new Parameters(PathElement.pathElement(ElytronDescriptionConstants.FILESYSTEM_REALM),
+                ElytronExtension.getResourceDescriptionResolver(ElytronDescriptionConstants.FILESYSTEM_REALM))
+            .setAddHandler(ADD)
+            .setRemoveHandler(REMOVE)
+            .setAddRestartLevel(OperationEntry.Flag.RESTART_RESOURCE_SERVICES)
+            .setRemoveRestartLevel(OperationEntry.Flag.RESTART_RESOURCE_SERVICES));
+    }
+
+    @Override
+    public void registerAttributes(ManagementResourceRegistration resourceRegistration) {
+        OperationStepHandler handler = new WriteAttributeHandler();
+        for (AttributeDefinition attr : ATTRIBUTES) {
+            resourceRegistration.registerReadWriteAttribute(attr, null, handler);
+        }
+    }
+
+    private static class RealmAddHandler extends AbstractAddStepHandler {
+
+        private RealmAddHandler() {
+            super(SECURITY_REALM_RUNTIME_CAPABILITY, ATTRIBUTES);
+        }
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model)
+                throws OperationFailedException {
+            ServiceTarget serviceTarget = context.getServiceTarget();
+            RuntimeCapability<Void> runtimeCapability = RuntimeCapability.fromBaseCapability(SECURITY_REALM_RUNTIME_CAPABILITY, context.getCurrentAddressValue());
+            ServiceName realmName = runtimeCapability.getCapabilityServiceName(SecurityRealm.class);
+            int levels = LEVELS.resolveModelAttribute(context, model).asInt();
+
+            String path = PATH.resolveModelAttribute(context, model).asString();
+            String relativeTo = RELATIVE_TO.resolveModelAttribute(context, model).asString();
+            String nameRewriter = model.hasDefined(NAME_REWRITER.getName()) ?
```

Comment:
Not a big deal but we have an asStringIfDefined method elsewhere to avoid needing to manually write this check all over the place.